[tensorexpr] Add support for aten::stack

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74077
* __->__ #73801

This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

huiguoo committed:
[tensorexpr] Add support for aten::stack

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.h:
```diff
@@ -17,7 +17,7 @@ TORCH_API void FuseTensorExprs(
     size_t min_group_size = 2,
     bool add_composed_op = false,
-    bool fuse_to_dynamic_shapes = false);
+    bool fuse_to_dynamic_shapes = true);
```
We shouldn't be changing the default as part of this PR i don't think

huiguoo replied:
Oh sorry it was a draft PR. Just updated it, and now it's ready for review! 

In this PR, aten::stack is not enabled for fusion yet. Will enable the fusion (w/wo dynamic shapes) in separate PRs.

eellison reviewed (commented):
Also we need shape analysis support for this for dynamic shapes..

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/misc.cpp:
```diff
@@ -642,6 +642,97 @@ Tensor computeCat(
 
+static bool checkStackInputShape(const std::vector<BufHandle>& bufList) {
+  if (bufList.size() == 0) {
+    throw std::runtime_error("Empty input list is passed to aten::stack");
```
Nit: let's be consistent in handling error cases. Two lines below we use `TORCH_INTERNAL_ASSERT` and here we're throwing an error.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/misc.cpp:
```diff
@@ -642,6 +642,97 @@ Tensor computeCat(
+
+        for (size_t ii = 1; ii < inputList.size(); ++ii) {
+          auto input = inputList[ii];
+          load = ifThenElse(
+              CompareSelect::make(
```
Why do we generate both `ifThenElse` and `CompareSelect`?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/misc.cpp:
```diff
@@ -642,6 +642,97 @@ Tensor computeCat(
+        int64_t dim_ = c10::get<int64_t>(argDim);
+        auto dim = normalizeAndCheckIndex(dim_, axes.size());
+        auto inputBuf = inputList[0];
+        std::vector<ExprHandle> newAxes(axes.begin(), axes.begin() + dim);
+        newAxes.insert(newAxes.end(), axes.begin() + dim + 1, axes.end());
```
This doesn't look right. First line makes
```
newAxes[:] = axes[:]
```
Second line then appends it with `axes[dim+1:]`.

That means that if we have 10-d tensors and dim=0 the stacked tensor will have rank 19. I think it should be 11-d instead, am I missing something?

Also, please update the comment to describe what new axis are created by this op and where they are inserted.

huiguoo replied:
The first line makes
```
newAxes = axes[:dim]
```
Overall, there will be N-1 elements in newAxes where N is the size of axes.

ZolotukhinM replied:
Ah, got it, it make sense. Maybe it'd be easier to grasp if we call `newAxes` as `outputAxes` and `axes` as `inputAxes` - currently "new" corresponds to inputs which is somewhat counterintuitive (even though I understand why it's done that way - we're deducing input indexes from output indexes).

ZolotukhinM commented on test/cpp/tensorexpr/test_ops.cpp:
```diff
@@ -41,3 +41,49 @@ TEST(Ops, Sum) {
+    buflist.push_back(a);
+    buflist.push_back(b);
+
+    auto at = at::rand(dims, at::kFloat);
+    auto bt = at::rand(dims, at::kFloat);
```
The test only checks stacking two tensors with equal dimensions. We should also check other cases, e.g.:
1) stacking just one tensor
2) stacking N>2 tensors
3) stacking tensors with different dimensions

It would be good to also have such tests in python, so that whenever opinfo covers `stack` we can use that too.

It might be also a good idea to have a test verifying IR that we generate.

huiguoo replied:
> The test only checks stacking two tensors with equal dimensions. We should also check other cases...

That's a good point! Will add more tests.

> It would be good to also have such tests in python, so that whenever opinfo covers `stack` we can use that too.

Not sure if I understand this. Are you suggesting to test above cases from Python instead of C++? We don't have many unit tests for ops right now. I only know aten::sum, aten::cat, and aten::conv. For these 3 ops, we put the unit tests in 3 files under `test/cpp/tensorexpr`. Maybe we should put them together in one file like `test_ops.cpp`. 

> It might be also a good idea to have a test verifying IR that we generate.

Agree. Will add it.

ZolotukhinM replied:
> Are you suggesting to test above cases from Python instead of C++? 

Having C++ tests is fine, but currently the most comprehensive coverage that we have is in python:
https://github.com/pytorch/pytorch/blob/122f8648ab8f872bd8ed970728c5ac894dce5300/test/test_jit_fuser_te.py#L2476

It leverages OpInfo, which is basically a collection of representative inputs for an op. What I suggested was to add a test for `stack` to those tests as well (if possible). If `stack` is not supported by OpInfo, then I think it still would make sense to write a python test manually, as most of our ops are currently tested from python rather than from C++.

huiguoo replied:
Thanks for the pointer!! Added such a test for `aten::stack` in PR #74077.

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

navahgar commented on torch/csrc/jit/tensorexpr/operators/misc.cpp:
```diff
@@ -642,6 +642,96 @@ Tensor computeCat(
+  return hasEmptyDims;
+}
+
+Tensor computeStack(
```
Can we add an implementation that lowers into multiple loops like in the case of `aten::cat`, as done in `computeCatWoConditionals`?

The conditionals based lowering does not give good perf on CPUs, mainly because we can't vectorize it. I think the conditionals based lowering still makes sense for CUDA.

huiguoo replied:
> The conditionals based lowering does not give good perf on CPUs, mainly because we can't vectorize it. 

`aten::stack` only uses `compareSelect` conditionals which can be vectorized. I'll see if the other implementation has better performance. Will add it if it is.

navahgar replied:
Please use some representative shapes from the models you have been working on, for perf analysis.

I still believe the multiple loops approach would be faster than conditionals because we will have both spacial and temporal cache locality for all inputs. It would be good to confirm with a perf analysis. 

If you want to do this as a followup, thats fine too.

huiguoo replied:
Sounds great! I'll first enable `aten::stack` in the fusion, and then do a perf analysis on real models for both approaches.

navahgar commented on torch/csrc/jit/tensorexpr/operators/misc.cpp:
```diff
@@ -642,6 +642,96 @@ Tensor computeCat(
+        }
+
+        int64_t dim_ = c10::get<int64_t>(argDim);
+        auto dim = normalizeAndCheckIndex(dim_, axes.size());
```
It looks like function errors out when `dim_ == axes.size()`. Isn't that a correct input for `stack`?

huiguoo replied:
The `axes` are for the output of `aten::stack`. For example, 
```
a = [1, 2]
b = [3, 4]
output = aten::stack([a, b], dim=dim_)
```
`axes.size()` equals to 3, and `dim_` should be a number between 0 to 2, or -1 to -3, right? It cannot be 3. We have such dim tests for `stack`.

navahgar replied:
Aah, I missed the point that `axes` is for the output of `stack`. Makes sense.

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

navahgar reviewed (commented):
LGTM

Please ensure you address all comments from Mikhail.

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

ZolotukhinM reviewed (approved):
Looks good! I think it's still a good idea to add a pyhton test though.

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

huiguoo commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -771,6 +771,47 @@ TEST_F(Kernel, OptimizeConditionals) {
+
+  const std::string& verification_pattern =
+      R"IR(
+# CHECK: for
+# CHECK-NEXT: for
+# CHECK-NEXT: for
+# CHECK-NEXT: for
+# CHECK-NEXT: aten_stack)IR";
```
Attempted to check the stmt IR in the following way,
```
CHECK-NEXT: aten_stack[Ramp(5ll * (j + 2ll * i), 1ll, 4)] = (Broadcast(j, 4))==(Broadcast(1ll, 4)) ? (ty_1[Ramp(5ll * i, 1ll, 4)]) : (tx_1[Ramp(5ll * i, 1ll, 4)]))
```
it failed one CI test (linux-xenial-py3.7-clang7-asan). With the configuration in this CI test, the vectorization failed thus the generated stmt is different. There are two ways to fix this,

1) implement a CHECK that can check patterns in a string such as ".*?.*:" for stack
2) save the original stmt before prepare_to_codegen in TEK for checking issues

Option 2 does not seem good as it introduces additional member vars in TEK only for testing issues; Option 1 can be implemented in a separate PR. Currently landing this PR with the original checks.

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

huiguoo committed:
Update on "[tensorexpr] Add support for aten::stack"


This PR adds the lowering function for aten::stack in NNC.
Differential Revision: [D34647822](https://our.internmc.facebook.com/intern/diff/D34647822)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34647822).

zengk95 commented:
@pytorchbot revert this

malfet commented:
Reverting, as it broke ASAN at Kernel.Stack test, see https://github.com/pytorch/pytorch/runs/5761874340?check_suite_focus=true
